src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseGradingService.java

```diff
@@ -728,57 +715,66 @@ private boolean createFeedbackForDuplicateTests(Result result, ProgrammingExerci
+            double score = calculateScore(programmingExercise, allTestCases, result, staticCodeAnalysisFeedbacks, applySubmissionPolicy);
             result.setScore(score, programmingExercise.getCourseViaExerciseGroupOrCourseMember());
         }
 
-        result.getFeedbacks().forEach(feedback -> {
-            if (feedback.getCredits() == null) {
-                feedback.setCredits(0D);
-            }
+        updateFeedbackCredits(result, allTestCases, programmingExercise.getMaxPoints());
+    }
+
+    private void updateFeedbackCredits(Result result, Set<ProgrammingExerciseTestCase> allTestCases, Double exerciseMaxPoints) {
+        double weightSum = calculateWeightSum(allTestCases);
+
+        // Set credits for successful test cases
+        allTestCases.stream().filter(testCase -> testCase.isSuccessful(result)).forEach(testCase -> {
+            double credits = calculatePointsForTestCase(result, exerciseMaxPoints, testCase, allTestCases.size(), weightSum);
+            setCreditsForTestCaseFeedback(result, testCase, credits);
         });
+
+        result.getFeedbacks().stream()//
+                .filter(feedback -> Objects.isNull(feedback.getCredits()))//
+                .forEach(feedback -> feedback.setCredits(0D));
     }
 
     /**
      * Calculates the score of automatic test cases for the given result with possible penalties applied.
-     * @param programmingExercise the result belongs to.
-     * @param allTests that should be considered in the score calculation.
-     * @param result for which a score should be calculated.
-     * @param successfulTestCases all test cases that passed for the submission.
+     *
+     * @param programmingExercise        the result belongs to.
+     * @param allTests                   that should be considered in the score calculation.
+     * @param result                     for which a score should be calculated.
      * @param staticCodeAnalysisFeedback that has been created for the submission.
-     * @param applySubmissionPolicy true, if penalties from submission policies should be applied.
-     * @return the final total score that should be given to the result.
+     * @param applySubmissionPolicy      true, if penalties from submission policies should be applied.
+     * @return the final total score in percent that should be given to the result.
      */
     private double calculateScore(final ProgrammingExercise programmingExercise, final Set<ProgrammingExerciseTestCase> allTests, final Result result,
-            final Set<ProgrammingExerciseTestCase> successfulTestCases, final List<Feedback> staticCodeAnalysisFeedback, boolean applySubmissionPolicy) {
-        if (successfulTestCases.isEmpty()) {
+            final List<Feedback> staticCodeAnalysisFeedback, boolean applySubmissionPolicy) {
+
+        // TODO: this might be redundant and covered in calculateSuccessfulTestPoints
```

Comment:
Could another more experienced developer affirm this for me? @b-fein @JohannesStoehr

Reply:
I think it might be redundant, the test case `shouldRecalculateScoreWithTestCaseBonusButNoExerciseBonus` for `result6` should cover that scenario. If that still passes and does not produce a crash somewhere, then it should be fine to remove the special case.